### ISSUE TITLE
[Core][TPU] Enhance Ray TPU utilities for heterogeneous training and inference workloads

### ISSUE THREAD

**Issue by ryanaoleary** [ray-project/ray#64071]
### Description

This can serve as a tracking issue for enhancements to the `ray.util.tpu` APIs. This utility library is meant to enable SPMD workloads on TPUs for both host-level (per TPU host) and device-level (per TPU chip or core) workloads. Some potential enhancements include:

- [x] Support single-host TPU node reservation in `SlicePlacementGroup`
  - https://github.com/ray-project/ray/pull/64079

- [x] For TPU multi-slice, `SlicePlacementGroup` exposes a `num_slices` field that dictates the number of TPU slices of a given `topology` to reserve atomically. The `SlicePlacementGroup` handle currently returns a single placement group over the `num_slices` TPU slices that are reserved. The use case is primarily for multi-slice training, but there are additional scenarios (RL, disaggregated serving, etc.) where users may want to reserve multiple slices at once, but run separate workloads per slice. To support this, we should introduce a `per_slice_pg` parameter to the `SlicePlacementGroup` class that updates the utility to return a list of PG handles rather than a single, broad PG.
  - https://github.com/ray-project/ray/pull/64072 

- [x] `SlicePlacementGroup`s should be able to sub-slice provisioned TPU slices (i.e. reserved by a RayCluster CR) at runtime. We can enhance the `TPUAcceleratorManager` to discover more topology information at Ray node init (host ID, sub-block ID, block ID, etc.) and add corresponding `ray.io/tpu-` labels. Then when a `SlicePlacementGroup` is created with an arbitrary `topology`- it should attempt to schedule utilizing these labels - even if the underlying GKE hardware provisioned is of a larger TPU topology/slice. This feature will enhance the flexibility of scheduling TPU workloads with Ray. 
  - https://github.com/ray-project/ray/pull/64578 

- [ ] The Slice PG API can utilize `topology_strategy` (https://github.com/ray-project/enhancements/pull/66) to achieve the desired gang scheduling behavior rather than the two-step PG scheduling (reserve TPU head -> retrieve slice name label -> reserve N TPU workers). We should help shepherd this proposal since this use-case maps well to TPUs, both for multi-host and multi-slice scheduling.
  - PR to enable hierarchical bundles and topology_strategy: https://github.com/ray-project/ray/pull/64950

- [ ] The `ray.util.tpu` APIs should be promoted to `beta` as they become more stable and better qualification testing is performed. Besides the unit and integration tests we run in this repo, a set of e2e tests exist [here](https://github.com/ai-on-gke/kuberay-tpu-webhook/tree/main/e2e) that are run on actual TPU hardware. Qualification testing on TPUs should be added to the Ray CI similar to these e2e tests.

- [ ] A utility to determine the `resources`, `bundle_label_selector`, `topology_strategy`, `scheduling_strategy` etc. to set given a `topology`, `accelerator_type` and `num_workers` (i.e. something like `get_accelerator_config`) would be useful to add - enabling users to pass these values to APIs like `serve.deployment` (https://docs.ray.io/en/latest/serve/api/doc/ray.serve.Deployment.html#ray-serve-deployment) to control scheduling behavior. This would be useful because configuring each of these fields for large workloads / TPU topologies can be error prone and confusing - this utility would simplify the deployment process for the user by deterministically calculating the desired values given just the high-level accelerator information (i.e. `topology`).

### Use case

The goal is for ML workloads (pre-training, inference, post-training) with frameworks like JAX, Torch, etc. to be easily configurable and ran through Ray on single-host, multi-host, and multi-slice TPUs.

These APIs should simplify the utilization of TPUs with JAX and Torch without requiring that users have a deep technical understanding of the exact underlying hardware. Beyond specifying a `topology` and `accelerator_type`, users should be able to write their ML workloads as though they're accelerator agnostic - leaving the complex distributed scheduling and initialization to the Ray backend.

**Comment by ryanaoleary** [comment]
cc: @andrewsykim @spencer-p
